### DISCUSSION TITLE
Help me make a Twitter plugin

### DISCUSSION THREAD

**Discussion by arye321** [streamlink/streamlink#5296] [General]
After realizing SpaceX's live stream on Twitter was 30s ahead I tried running it in Streamlink and found out theres no supporting plugin ...

I made most of the work on how to find the m3u8 link, just need some help converting it to a plugin file because I couldn't find suitable resources on how to make one.

this is the python code to get the m3u8:

    import requests
    from random import randint
    
    url = "https://twitter.com/i/broadcasts/1kvJpmapzboxE"
    
    liveID = url.rstrip('/').rsplit('/', 1)[-1]
    # liveID = "1kvJpmapzboxE"
    headers = {
        'authorization': 'Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA',
        'x-guest-token': str(randint(1,99999999)),
    }
    params = {
        'include_events': 'true',
        'ids': liveID,
        'broadcastVersionMap': '',
    }
    response = requests.get('https://twitter.com/i/api/1.1/broadcasts/show.json', params=params, headers=headers)
    
    
    media_key = response.json()['broadcasts'][liveID]['media_key']
    
    response = requests.get(
        f'https://twitter.com/i/api/1.1/live_video_stream/status/{media_key}',
          headers=headers,
    )
    
    print(response.json()["source"]["location"])
    

What I do is send the link ID (1kvJpmapzboxE) to

  
    https://twitter.com/i/api/1.1/broadcasts/show.json

this returns a media_key which i send to 

    https://twitter.com/i/api/1.1/live_video_stream/status/{media_key}

 
this prints the m3u8 link

    https://twitter.com/i/live_video_stream/authorized_status/1648157153457483777/LIVE_PUBLIC/Ft9uMuHaMAE4Iao?url=https://prod-fastly-ap-southeast-1.video.pscp.tv/Transcoding/v1/hls/WkSJYlqJeu3ifMNPU5v8wESwT4r0UwXxyht9XdUQZFtsFXVLwyeyoPCfXvQ3XU8SJiOgAWLijnJfezUT85GTRw/non_transcode/ap-southeast-1/periscope-replay-direct-prod-ap-southeast-1-public/tw_master_dynamic_highlatency.m3u8?type%3Dlive&ctx=AAAAEIDbBhuH0rmssvj6UXK8l1MelmXca8siO34rJXEn3uyGxA4MfS1mf3gUOUy7&evt=28_1648157153457483777&exp=1681793951653&checksum=yf_CSwwSud8rtKCC

which works if I paste it in VLC

if you know how to make a Streamlink plugin, this should be pretty trivial to continue from here, I think.

**edit:** ok i've made some progress, this works as a plugin now:

    
```python 

from streamlink.plugin import Plugin
from streamlink.plugin.api import validate
from streamlink.stream import HLSStream
import re
from random import randint


class Twitter(Plugin):
    url_re = re.compile(r"https?://twitter\.com/i/broadcasts/\w+")

    @classmethod
    def can_handle_url(cls, url):
        return cls.url_re.match(url) is not None

    def _get_streams(self):
        liveID = self.url.rstrip("/").rsplit("/", 1)[-1]
        headers = {
            "authorization": "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA",
            "x-guest-token": str(randint(1, 99999999)),
        }
        params = {
            "include_events": "true",
            "ids": liveID,
            "broadcastVersionMap": "",
        }
        res = self.session.http.get(
            "https://twitter.com/i/api/1.1/broadcasts/show.json",
            headers=headers,
            params=params,
        )
        jsonres = self.session.http.json(res)
        media_key = jsonres["broadcasts"][liveID]["media_key"]

        res = self.session.http.get(
            f"https://twitter.com/i/api/1.1/live_video_stream/status/{media_key}",
            headers=headers,
        )
        jsonres = self.session.http.json(res)
        m3u8 = jsonres["source"]["location"]

        return HLSStream.parse_variant_playlist(self.session, m3u8)


__plugin__ = Twitter

```

**Comment by bastimeyer** [comment]
> ```python
>     @classmethod
>     def can_handle_url(cls, url):
>         return cls.url_re.match(url) is not None
> ```

This is a plugin API which has been deprecated a long time ago and which will be removed at some point soon.
https://streamlink.github.io/deprecations.html#plugin-can-handle-url-and-plugin-priority

In regards to your actual plugin implementation, this will raise various different errors when trying to access URLs without a live stream, which won't be handled by Streamlink and will result in a stack trace being printed by the Python interpreter.

You're also hardcoding a bearer token, which is probably not a wise idea.

**Reply by arye321** [reply]
Thanks but this is for personal use, if it brakes ill fix it. 

the bearer is very public and been used for years, if you look it up you'll see there are repos that used it 4 years ago

**Reply by WolfganP** [reply]
Thx a lot @arye321 for this plugin attempt for twitter broadcasts. For me it conflicts with generic plugin, so additional troubleshooting needed from my side to allow them work at the same time.

I'll try to convert it to the new URL matching API as well.

**Reply by arye321** [reply]
```
from streamlink.plugin import Plugin, pluginmatcher
from streamlink.plugin.api import validate
from streamlink.stream import HLSStream
import re
from random import randint


@pluginmatcher(re.compile(r"https?://(?:\S+\.)?twitter.com/i/broadcasts/(?P<id>\w+)"))
class Twitter(Plugin):
    def _get_streams(self):
        # liveID = self.url.rstrip("/").rsplit("/", 1)[-1]
        liveID = self.match.group("id")
        headers = {
            "authorization": "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA",
            "x-guest-token": str(randint(1, 99999999)),
        }
        params = {
            "include_events": "true",
            "ids": liveID,
            "broadcastVersionMap": "",
        }
        res = self.session.http.get(
            "https://twitter.com/i/api/1.1/broadcasts/show.json",
            headers=headers,
            params=params,
        )
        jsonres = self.session.http.json(res)
        media_key = jsonres["broadcasts"][liveID]["media_key"]

        res = self.session.http.get(
            f"https://twitter.com/i/api/1.1/live_video_stream/status/{media_key}",
            headers=headers,
        )
        jsonres = self.session.http.json(res)
        m3u8 = jsonres["source"]["location"]

        # video_url = self.session.http.json(res)["data"]["video_url"]
        return HLSStream.parse_variant_playlist(self.session, m3u8)


__plugin__ = Twitter


```

**Reply by WolfganP** [reply]
Thanks! You won me to it. I just added priority to it to avoid some previous conflicts.

`from streamlink.plugin import Plugin, pluginmatcher, HIGH_PRIORITY`

**Comment by mkbloke** [comment]
Glad you got it working.  I think they also stream on YT, for which there is a Streamlink plugin, via: https://www.youtube.com/@SpaceX/streams

**Reply by arye321** [reply]
Thanks, as I mentioned, Twitters stream is 30 seconds ahead

**Comment by arye321** [comment]
updated for X:

```python
from streamlink.plugin import Plugin, pluginmatcher
from streamlink.plugin.api import validate
from streamlink.stream import HLSStream
import re
from random import randint


@pluginmatcher(re.compile(r"https?://(?:\S+\.)?(?:twitter|x)\.com/i/broadcasts/(?P<id>\w+)"))
class Twitter(Plugin):
    def _get_streams(self):
        # liveID = self.url.rstrip("/").rsplit("/", 1)[-1]
        liveID = self.match.group("id")
        headers = {
            "authorization": "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA",
            "x-guest-token": str(randint(1, 99999999)),
        }
        params = {
            "include_events": "true",
            "ids": liveID,
            "broadcastVersionMap": "",
        }
        res = self.session.http.get(
            "https://twitter.com/i/api/1.1/broadcasts/show.json",
            headers=headers,
            params=params,
        )
        jsonres = self.session.http.json(res)
        media_key = jsonres["broadcasts"][liveID]["media_key"]

        res = self.session.http.get(
            f"https://twitter.com/i/api/1.1/live_video_stream/status/{media_key}",
            headers=headers,
        )
        jsonres = self.session.http.json(res)
        m3u8 = jsonres["source"]["location"]

        return HLSStream.parse_variant_playlist(self.session, m3u8)


__plugin__ = Twitter
```

**Reply by WolfganP** [reply]
SpaceX stream works like a charm. Thx a lot for the update!
